### PR TITLE
[#137] Expire session token locally

### LINKED ISSUE #137: Expire session tokens locally in pool
https://github.com/nspcc-dev/neofs-sdk-go/issues/126 has been solved with a simple solution from #129.
We can go further and implement more complex solution that reduces amount of new session tokens created by `pool`. Solution proposed by @realloc 

In #129 we couldn't use gRPC state monitor because it doesn't cover token expiration case. In other words, `pool` **can** renew session token when remote storage node has been restarted (its state changes from `READY`), but it **can't** renew session token if it was expired. 

To address this case, `pool` can monitor and expire session tokens from the cache manually. To do that `pool` can use passive and active epoch observing.

Passive: epoch value is stored in every `ResponseMetaHeader`, so each successful request updates local epoch number.
Active: during healthcheck, `pool` sends `NodeInfo` request, so `ResponseMetaHeader` with epoch value can be processed there as well.

With combination of these two observations, it is possible to expire session tokens locally.

### PR DESCRIPTION
**Pull request by KirillovDenis**
Should be used along with https://github.com/nspcc-dev/neofs-node/issues/1189

Blocked by https://github.com/nspcc-dev/neofs-node/issues/1191

closes #137

### CODE REVIEW

File: pool/cache.go [hunk 1]
```diff
@@ -61,3 +68,15 @@ func (c *SessionCache) DeleteByPrefix(prefix string) {
 		}
 	}
 }
+
+func (c *SessionCache) UpdateEpoch(newEpoch uint64) {
```
Comment (fyrchik) [inline pool/cache.go:72]: Unused in this commit, should it be used from elsewhere?

File: client/session.go [hunk 2]
```diff
@@ -48,6 +50,15 @@ func (x ResSessionCreate) PublicKey() []byte {
 	return x.sessionKey
 }
 
+func (x *ResSessionCreate) setExp(exp uint64) {
+	x.exp = exp
+}
+
+// Exp returns epoch number of the token expiration.
+func (x ResSessionCreate) Exp() uint64 {
```
Comment (fyrchik) [inline client/session.go:58]: What about a more descriptive name for a public method? `Expiration`?

File: pool/cache.go (hunk 1 above)
Comment (KirillovDenis) [reply pool/cache.go:72]: This is used in https://github.com/nspcc-dev/neofs-sdk-go/pull/141/commits/56aa8e159cfc6c6ffe0a8fa9e2db41cc7996eb83

File: client/response.go [hunk 3]
```diff
@@ -18,3 +20,8 @@ type responseV2 interface {
 func (x ResponseMetaInfo) ResponderKey() []byte {
 	return x.key
 }
+
+// Epoch return current epoch.
```
Comment (cthulhu-rider) [inline client/response.go:24]:
```suggestion
// Epoch returns local NeoFS epoch of the server.
```

File: client/session.go [hunk 4]
```diff
@@ -50,6 +52,15 @@ func (x ResSessionCreate) PublicKey() []byte {
 	return x.sessionKey
 }
 
+func (x *ResSessionCreate) setExp(exp uint64) {
+	x.exp = exp
+}
+
+// Expiration returns epoch number of the token expiration.
+func (x ResSessionCreate) Expiration() uint64 {
```
Comment (cthulhu-rider) [inline client/session.go:60]: Why do we need this method? Value doesn't differ with `PrmSessionCreate.SetExp`.

File: pool/cache.go [hunk 5]
```diff
@@ -61,3 +68,15 @@ func (c *sessionCache) DeleteByPrefix(prefix string) {
 		}
 	}
 }
+
+func (c *sessionCache) UpdateEpoch(newEpoch uint64) {
```
Comment (cthulhu-rider) [inline pool/cache.go:72]:
```suggestion
func (c *sessionCache) updateEpoch(newEpoch uint64) {
```

**Review by cthulhu-rider — APPROVED** [review]

**Review by alexvanin — APPROVED** [review]

**Pull request merged**